### PR TITLE
[Group 2] Enable nullable annotations for `Microsoft.Extensions.Configuration.Abstractions`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to #43605, #54012

Annotated according to:
- https://github.com/dotnet/runtime/pull/57395

### CODE REVIEW

**Comment by maryamariyan** [comment]
Thanks for the PRs @maxkoshevoi, the nullable annotations effort for Microsoft.Extensions.* is targeted for 7.0, and since this week we're focusing on 6.0 bugs, there might be some delays with the reviews.

**Comment by maxkoshevoi** [comment]
I hoped, at least `Configuration.Abstractions` can still make it into 6.0.. Got it, thanks.

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs [hunk 1]
```diff
@@ -26,16 +26,17 @@ public static partial class ConfigurationPath
         public static readonly string KeyDelimiter;
         public static string Combine(System.Collections.Generic.IEnumerable<string> pathSegments) { throw null; }
         public static string Combine(params string[] pathSegments) { throw null; }
-        public static string GetParentPath(string path) { throw null; }
-        public static string GetSectionKey(string path) { throw null; }
+        public static string? GetParentPath(string? path) { throw null; }
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs:29]: `path` being nullable sounds counter-intuitive to me. Is this documented? If not consider making this non-null

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.csproj [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <Nullable>enable</Nullable>
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="Microsoft.Extensions.Configuration.Abstractions.cs" />
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Primitives\ref\Microsoft.Extensions.Primitives.csproj" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.csproj:12]: why is this needed after nullable changes?

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs [hunk 3]
```diff
@@ -65,7 +67,7 @@ public static string GetSectionKey(string path)
         /// </summary>
         /// <param name="path">The path.</param>
         /// <returns>The original path minus the last individual segment found in it. Null if the original path corresponds to a top level node.</returns>
-        public static string GetParentPath(string path)
+        public static string? GetParentPath(string? path)
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs:70]: honestly I think `path` should be marked as `non-null` the doc suggests it should be a path also it's not intuitive what will happen when it's null. I do not feel strongly about this so fine with me if you think we should keep it as is

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs [hunk 4]
```diff
@@ -17,14 +18,14 @@ public interface IConfigurationProvider
         /// <param name="key">The key.</param>
         /// <param name="value">The value.</param>
         /// <returns><c>True</c> if a value for the specified key was found, otherwise <c>false</c>.</returns>
-        bool TryGet(string key, out string value);
+        bool TryGet(string key, [MaybeNullWhen(false)] out string value);
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs:21]: I think it would be more consistent (i.e. with dictionary) for this to be nullable with `NotNullWhen(true)`

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs [hunk 5]
```diff
@@ -45,6 +46,6 @@ public interface IConfigurationProvider
         /// <param name="earlierKeys">The child keys returned by the preceding providers for the same parent path.</param>
         /// <param name="parentPath">The parent path.</param>
         /// <returns>The child keys.</returns>
-        IEnumerable<string> GetChildKeys(IEnumerable<string> earlierKeys, string parentPath);
+        IEnumerable<string> GetChildKeys(IEnumerable<string> earlierKeys, string? parentPath);
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs:49]: same comment about path being null

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationSection.cs [hunk 6]
```diff
@@ -21,6 +21,6 @@ public interface IConfigurationSection : IConfiguration
         /// <summary>
         /// Gets or sets the section value.
         /// </summary>
-        string Value { get; set; }
+        string? Value { get; set; }
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationSection.cs:24]: if you decide to keep path as nullable arg in other places should this also accept nullable path?

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/Microsoft.Extensions.Configuration.Abstractions.csproj [hunk 7]
```diff
@@ -1,8 +1,11 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/Microsoft.Extensions.Configuration.Abstractions.csproj:4]: why are these changes needed?

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs (hunk 4 above)
Comment (krwq) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs:21]: also since you allow assigning null in the `Set` should this just always be nullable? (assuming assigning false doesn't mean delete)

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/Microsoft.Extensions.Configuration.Abstractions.csproj (hunk 7 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/Microsoft.Extensions.Configuration.Abstractions.csproj:4]: `netstandard2.0` has very limited nullable support (if any, I don't remember). NetCore target framework is needed for proper nullable annotation (some nullable attributes were just added in net5, for example).

- Here's the comment asking to add `$(NetCoreAppCurrent)` along with nullable annotations: https://github.com/dotnet/runtime/issues/57200#issuecomment-897710642
- And here's the issue for that: https://github.com/dotnet/runtime/issues/54012 (also mentioned in PRs description)

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.csproj (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.csproj:12]: `$(NetCoreAppCurrent)` target was added here since it was also added for `src`. See https://github.com/dotnet/runtime/pull/57401#discussion_r709931489

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs (hunk 3 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs:70]: I also don't have a strong opinion. One benefit of making `path` nullable is that user don't need additional null-check in order to call this method (if their variable may be null).

I didn't find any calls to this method here or in aspnet, so don't know how this method it used.

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs (hunk 4 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationProvider.cs:21]: Good catch! Configuration values can be `null`, so `TryGet` can return `null` even if value exists.

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationSection.cs (hunk 6 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationSection.cs:24]: I don't think so. `Path` will never be set to `null`. All sections have a path.

Also all usages don't expect null in `Path`:
- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationRootExtensions.cs#L28
- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs#L57
- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs#L64

Regarding leaving `path` parameters us null. I'm not objecting removing `?`. My only concern is that it will be harder to call those methods by requiring additional null-check on their side.

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.cs:29]: Let's continue discussion in https://github.com/dotnet/runtime/pull/57401#discussion_r709917799

**Review by eerhardt — APPROVED** [review]
I think these changes look good. Thanks @maxkoshevoi.
